src/device_info.sh

```diff
@@ -0,0 +1,261 @@
+  case "$target" in
+    2) # LOCAL_TARGET
+      cpu_model=$(cmd_manager "$flag" "$cmd_model")
+      cpu_frequency=$(cmd_manager "$flag" "$cmd_frequency")
+      ;;
+  esac
+
+  device_info_data['cpu_model']="$cpu_model"
+
+  if [[ "$flag" == 'TEST_MODE' ]]; then
+    echo "$cpu_model"
+    echo "$cpu_frequency"
+    return 0
+  fi
+
+  cpu_currently=$(echo "$cpu_frequency" | grep 'CPU MHz')
+  cpu_max=$(echo "$cpu_frequency" | grep 'CPU max MHz')
+  cpu_min=$(echo "$cpu_frequency" | grep 'CPU min MHz')
+
+  cpu_currently=${cpu_currently//[!0-9,.]/}
+  cpu_max=${cpu_max//[!0-9,.]/}
+  cpu_min=${cpu_min//[!0-9,.]/}
+
+  device_info_data['cpu_currently']="$cpu_currently"
+  device_info_data['cpu_max']="$cpu_max"
+  device_info_data['cpu_min']="$cpu_min"
+}
+
+# This function populates the values from the size and fs (filesystem) key of
+# the device_info_data variable.
+#
+# @target Target machine
+function get_disk()
+{
+  local target="${device_options['target']}"
+  local flag="$1"
+  local cmd
+  local info
+  local size
+  local mount
+  local fs
+
+  flag=${flag:-'SILENT'}
+  cmd="df -h / | tail -n 1 | tr -s ' '"
+  case "$target" in
+    2) # LOCAL_TARGET
+      info=$(cmd_manager "$flag" "$cmd")
+      ;;
+  esac
+
+  if [[ "$flag" == 'TEST_MODE' ]]; then
+    echo "$info"
+    return 0
+  fi
+
+  fs="$(echo "$info" | cut -d' ' -f1)"
+  size="$(echo "$info" | cut -d' ' -f2)"
+  mount="$(echo "$info" | cut -d' ' -f6)"
+
+  device_info_data['disk_size']="$size"
+  device_info_data['root_path']="$fs"
+  device_info_data['fs_mount']="$mount"
+}
+
+# This function populates the os variable from the device_info_data variable.
+#
+# @target Target machine
+function get_os()
+{
+  local target="${device_options['target']}"
+  local os
+
+  case "$target" in
+    2) # LOCAL_TARGET
+      os=$(detect_distro '/')
+      ;;
+  esac
+
+  device_info_data['os']="$os"
+}
+
+# This function populates the gpu associative array with the vendor and
+# fetchable memory from each GPU found in the target machine.
+function get_gpu()
+{
+  local target="${device_options['target']}"
+  local flag="$1"
+  local cmd
+  local info
+  local gpu_info
+
+  cmd="lspci | grep -e VGA -e 3D | cut -d' ' -f1"
+  case "$target" in
+    2) # LOCAL_TARGET
+      info=$(cmd_manager "$flag" "$cmd")
+      ;;
+    3) # REMOTE_TARGET
+      info=$(cmd_remotely "$cmd" "$flag" "$ip" "$port")
+      ;;
+  esac
+
+  for g in $info; do
+    gpu_info=$(lspci -v -s "$g")
+    vendor=$(echo "$gpu_info" | head -n 1 | sed -r 's/.+controller: ([^ ]+).+/\1/')
+    memory=$(echo "$gpu_info" | sed -n -r '/Memory.+, prefetchable/s/.+\[size=(.+)\]/\1/p' | tr '\n' ',' | sed 's/.$//')
+    gpus["$g"]="$vendor,$memory"
+  done
+}
+
+# This function calls other functions to populate the device_info_data variable
+# with the data related to the hardware from the target machine.
+#
+# @target Target machine
+function learn_device()
```

Comment:
This function depends on `device_info`, which is not ideal. I want to use this function in the deploy context, rewrite it in a way that I can use `include devices.sh` in the `deploy.sh`, and just call this function to collect that hardware data that I need.